File: src/lib/book.ts
import { inferVenue } from "./catalog";
import type {
  Fill,
  Market,
  Order,
  PortfolioSnapshot,
  Position,
  ProposedTrade,
  Side,
} from "./types";
import { uid } from "./utils";

export const FEE_BPS = 6;

export function markPrice(markets: Market[], symbol: string): number {
  return markets.find((m) => m.symbol === symbol)?.price ?? 0;
}

export function positionPnl(pos: Position, px: number): number {
  const dir = pos.side === "long" ? 1 : -1;
  return (px - pos.avgPrice) * pos.qty * dir;
}

export function equityOf(
  cash: number,
  positions: Position[],
  markets: Market[],
): number {
  let eq = cash;
  for (const p of positions) {
    const px = markPrice(markets, p.symbol);
    if (!px) continue;
    eq += positionPnl(p, px);
  }
  return eq;
}

export function notional(pos: Position, px: number): number {
  return Math.abs(pos.qty) * px;
}

export type BookState = {
  cash: number;
  positions: Position[];
  orders: Order[];
  fills: Fill[];
  killSwitch: boolean;
};

export function emptyBook(): BookState {
  return {
    cash: 100_000,
    positions: [],
    orders: [],
    fills: [],
    killSwitch: false,
  };
}

export function applyTrade(
  book: BookState,
  markets: Market[],
  trade: ProposedTrade,
): { book: BookState; order: Order; fill?: Fill; error?: string; realized?: number } {
  if (book.killSwitch) {
    return { book, order: ghostOrder(trade, "cancelled"), error: "Kill switch is on." };
  }
  const market = markets.find((m) => m.symbol === trade.symbol);
  if (!market) {
    return { book, order: ghostOrder(trade, "cancelled"), error: `Unknown symbol ${trade.symbol}` };
  }
  const px = trade.type === "limit" && trade.price ? trade.price : market.price;
  if (!(px > 0) || !(trade.qty > 0)) {
    return { book, order: ghostOrder(trade, "cancelled"), error: "Invalid price or size." };
  }
  const leverage = Math.max(1, Math.min(trade.leverage ?? 1, market.venue === "perp" ? 20 : 1));
  const cost = (px * trade.qty) / leverage;
  const fee = px * trade.qty * (FEE_BPS / 10_000);
  const venue = market.venue;

  if (trade.type === "limit" && trade.price && !crosses(trade.side, trade.price, market.price)) {
    const order: Order = {
      id: uid("ord"),
      symbol: trade.symbol,
      venue,
      side: trade.side,
      type: "limit",
      qty: trade.qty,
      price: trade.price,
      leverage,
      status: "open",
      createdAt: Date.now(),
    };
    return { book: { ...book, orders: [order, ...book.orders] }, order };
  }

  if (trade.side === "buy") {
    if (book.cash < cost + fee) {
      return {
        book,
        order: ghostOrder(trade, "cancelled"),
        error: `Need ${ (cost + fee).toFixed(2) } USD, have ${book.cash.toFixed(2)}.`,
      };
    }
    const next = openOrAdd(book, trade.symbol, venue, "long", trade.qty, px, leverage);
    const order = filledOrder(trade, venue, px, leverage);
    const fill = makeFill(order, px, fee);
    return {
      book: {
        ...next,
        cash: next.cash - cost - fee,
        orders: [order, ...next.orders],
        fills: [fill, ...next.fills],
      },
      order,
      fill,
    };
  }

  // sell: reduce long or open short (perps only)
  const existing = book.positions.find((p) => p.symbol === trade.symbol);
  if (existing && existing.side === "long") {
    const closeQty = Math.min(existing.qty, trade.qty);
    const proceeds = (px * closeQty) / existing.leverage;
    const realized = (px - existing.avgPrice) * closeQty;
    const leftover = existing.qty - closeQty;
    const positions =
      leftover <= 1e-10
        ? book.positions.filter((p) => p.id !== existing.id)
        : book.positions.map((p) =>
            p.id === existing.id ? { ...p, qty: leftover } : p,
          );
    const order = filledOrder(trade, venue, px, existing.leverage);
    const fill = makeFill(order, px, fee);
    let next: BookState = {
      ...book,
      cash: book.cash + proceeds + realized - fee,
      positions,
      orders: [order, ...book.orders],
      fills: [fill, ...book.fills],
    };
    const remainder = trade.qty - closeQty;
    if (remainder > 1e-10) {
      if (venue !== "perp") {
        return { book: next, order, fill };
      }
      const shorted = applyTrade(
        next,
        markets,
        { ...trade, qty: remainder, side: "sell", reason: trade.reason },
      );
      return shorted;
    }
    return { book: next, order, fill, realized };
  }

  if (venue !== "perp" && venue !== "predict") {
    return {
      book,
      order: ghostOrder(trade, "cancelled"),
      error: "Shorting is only enabled on perps and prediction markets.",
    };
  }
  if (book.cash < cost + fee) {
    return {
      book,
      order: ghostOrder(trade, "cancelled"),
      error: `Need ${ (cost + fee).toFixed(2) } USD margin, have ${book.cash.toFixed(2)}.`,
    };
  }
  const next = openOrAdd(book, trade.symbol, venue, "short", trade.qty, px, leverage);
  const order = filledOrder(trade, venue, px, leverage);
  const fill = makeFill(order, px, fee);
  return {
    book: {
      ...next,
      cash: next.cash - cost - fee,
      orders: [order, ...next.orders],
      fills: [fill, ...next.fills],
    },
    order,
    fill,
  };
}

function crosses(side: Side, limit: number, mark: number): boolean {
  return side === "buy" ? limit >= mark : limit <= mark;
}

function openOrAdd(
  book: BookState,
  symbol: string,
  venue: Position["venue"],
  side: Position["side"],
  qty: number,
  px: number,
  leverage: number,
): BookState {
  const existing = book.positions.find((p) => p.symbol === symbol && p.side === side);
  if (!existing) {
    const pos: Position = {
      id: uid("pos"),
      symbol,
      venue,
      side,
      qty,
      avgPrice: px,
      leverage,
      openedAt: Date.now(),
    };
    return { ...book, positions: [pos, ...book.positions] };
  }
  const total = existing.qty + qty;
  const avg = (existing.avgPrice * existing.qty + px * qty) / total;
  return {
    ...book,
    positions: book.positions.map((p) =>
      p.id === existing.id ? { ...p, qty: total, avgPrice: avg } : p,
    ),
  };
}

function ghostOrder(trade: ProposedTrade, status: Order["status"]): Order {
  return {
    id: uid("ord"),
    symbol: trade.symbol,
    venue: inferVenue(trade.symbol),
    side: trade.side,
    type: trade.type,
    qty: trade.qty,
    price: trade.price ?? 0,
    leverage: trade.leverage ?? 1,
    status,
    createdAt: Date.now(),
  };
}

function filledOrder(
  trade: ProposedTrade,
  venue: Order["venue"],
  px: number,
  leverage: number,
): Order {
  const now = Date.now();
  return {
    id: uid("ord"),
    symbol: trade.symbol,
    venue,
    side: trade.side,
    type: trade.type,
    qty: trade.qty,
    price: px,
    leverage,
    status: "filled",
    createdAt: now,
    filledAt: now,
  };
}

function makeFill(order: Order, px: number, fee: number): Fill {
  return {
    id: uid("fil"),
    orderId: order.id,
    symbol: order.symbol,
    side: order.side,
    qty: order.qty,
    price: px,
    fee,
    createdAt: Date.now(),
  };
}

export function snapshotOf(book: BookState, markets: Market[]): PortfolioSnapshot {
  return {
    cash: book.cash,
    equity: equityOf(book.cash, book.positions, markets),
    positions: book.positions,
    openOrders: book.orders.filter((o) => o.status === "open"),
    dayPnl: equityOf(book.cash, book.positions, markets) - 100_000,
    killSwitch: book.killSwitch,
  };
}

export function tryParseTrades(text: string): ProposedTrade[] {
  const out: ProposedTrade[] = [];
  const fence = /```(?:trade|json)\s*([\s\S]*?)```/gi;
  let m: RegExpExecArray | null;
  while ((m = fence.exec(text))) {
    try {
      const raw = JSON.parse(m[1]!) as unknown;
      const items = Array.isArray(raw) ? raw : [raw];
      for (const item of items) {
        const t = coerceTrade(item);
        if (t) out.push(t);
      }
    } catch {
      /* ignore */
    }
  }
  return out;
}

function coerceTrade(raw: unknown): ProposedTrade | null {
  if (!raw || typeof raw !== "object") return null;
  const o = raw as Record<string, unknown>;
  const symbol = String(o.symbol ?? "").toUpperCase();
  const side = o.side === "sell" ? "sell" : o.side === "buy" ? "buy" : null;
  if (!symbol || !side) return null;
  const qty = Number(o.qty ?? o.size ?? o.quantity);
  if (!(qty > 0)) return null;
  return {
    symbol,
    side,
    type: o.type === "limit" ? "limit" : "market",
    qty,
    price: typeof o.price === "number" ? o.price : undefined,
    leverage: typeof o.leverage === "number" ? o.leverage : undefined,
    reason: String(o.reason ?? o.thesis ?? ""),
  };
}

File: desktop/src/book.rs
use serde::{Deserialize, Serialize};

#[derive(Clone, Debug, Serialize, Deserialize)]
pub struct Position {
    pub symbol: String,
    pub side: String,
    pub qty: f64,
    pub avg: f64,
}

#[derive(Clone, Debug, Serialize, Deserialize)]
pub struct Fill {
    pub symbol: String,
    pub side: String,
    pub qty: f64,
    pub price: f64,
    pub realized: Option<f64>,
}

#[derive(Clone, Debug, Serialize, Deserialize)]
pub struct Book {
    pub cash: f64,
    pub kill: bool,
    pub positions: Vec<Position>,
    pub fills: Vec<Fill>,
}

impl Default for Book {
    fn default() -> Self {
        Self {
            cash: 100_000.0,
            kill: false,
            positions: vec![],
            fills: vec![],
        }
    }
}

impl Book {
    pub fn equity(&self, marks: &[(String, f64)]) -> f64 {
        let mut eq = self.cash;
        for p in &self.positions {
            let px = marks
                .iter()
                .find(|(s, _)| s == &p.symbol)
                .map(|(_, v)| *v)
                .unwrap_or(p.avg);
            let signed = if p.side == "short" { -1.0 } else { 1.0 };
            eq += signed * (px - p.avg) * p.qty + p.avg * p.qty;
        }
        eq
    }

    pub fn apply(&mut self, symbol: &str, side: &str, qty: f64, px: f64) -> Result<Option<f64>, String> {
        if self.kill {
            return Err("Kill switch is on.".into());
        }
        if !(px > 0.0 && qty > 0.0) {
            return Err("Invalid size.".into());
        }
        if side == "buy" {
            let cost = px * qty;
            if self.cash < cost {
                return Err(format!("Need {cost:.0}, have {:.0}", self.cash));
            }
            self.cash -= cost;
            if let Some(p) = self.positions.iter_mut().find(|p| p.symbol == symbol && p.side == "long") {
                let total = p.qty + qty;
                p.avg = (p.avg * p.qty + px * qty) / total;
                p.qty = total;
            } else {
                self.positions.push(Position {
                    symbol: symbol.into(),
                    side: "long".into(),
                    qty,
                    avg: px,
                });
            }
            self.fills.insert(0, Fill {
                symbol: symbol.into(),
                side: "buy".into(),
                qty,
                price: px,
                realized: None,
            });
            return Ok(None);
        }

        if let Some(idx) = self.positions.iter().position(|p| p.symbol == symbol && p.side == "long") {
            let p = &self.positions[idx];
            let close = qty.min(p.qty);
            let realized = (px - p.avg) * close;
            self.cash += px * close;
            if close >= p.qty - 1e-10 {
                self.positions.remove(idx);
            } else {
                self.positions[idx].qty -= close;
            }
            self.fills.insert(0, Fill {
                symbol: symbol.into(),
                side: "sell".into(),
                qty: close,
                price: px,
                realized: Some(realized),
            });
            return Ok(Some(realized));
        }

        let cost = px * qty;
        if self.cash < cost {
            return Err(format!("Need {cost:.0} margin, have {:.0}", self.cash));
        }
        self.cash -= cost;
        if let Some(p) = self.positions.iter_mut().find(|p| p.symbol == symbol && p.side == "short") {
            let total = p.qty + qty;
            p.avg = (p.avg * p.qty + px * qty) / total;
            p.qty = total;
        } else {
            self.positions.push(Position {
                symbol: symbol.into(),
                side: "short".into(),
                qty,
                avg: px,
            });
        }
        self.fills.insert(0, Fill {
            symbol: symbol.into(),
            side: "sell".into(),
            qty,
            price: px,
            realized: None,
        });
        Ok(None)
    }
}
